loader/reset.go

```diff
@@ -144,20 +145,46 @@ func (p *ResetProcessor) applyNullOverrides(target any, path tree.Path) error {
 }
 
 func (p *ResetProcessor) checkForCycle(node *yaml.Node, path tree.Path) error {
-	if p.visitedNodes == nil {
-		p.visitedNodes = make(map[*yaml.Node]string)
-	}
+	paths := p.visitedNodes[node]
+	pathStr := path.String()
 
-	// Check for cycle by seeing if the node has already been visited at this path
-	if previousPath, found := p.visitedNodes[node]; found {
-		// If the current node has been visited, we have a cycle if the previous path is a prefix
-		if strings.HasPrefix(path.String(), strings.TrimRight(previousPath, "<<")) {
-			return fmt.Errorf("cycle detected at path: %s", previousPath)
+	for _, prevPath := range paths {
+		// If we're visiting the exact same path, it's not a cycle
+		if pathStr == prevPath {
+			continue
 		}
-	}
 
-	// Mark the current node as visited
-	p.visitedNodes[node] = path.String()
+		// If either path is using a merge key, it's legitimate YAML merging
+		if strings.Contains(prevPath, "<<") || strings.Contains(pathStr, "<<") {
+			continue
+		}
 
+		// Only consider it a cycle if one path is contained within the other
+		// and they're not in different service definitions
+		if (strings.HasPrefix(pathStr, prevPath+".") ||
+			strings.HasPrefix(prevPath, pathStr+".")) &&
+			!areInDifferentServices(pathStr, prevPath) {
+			return fmt.Errorf("cycle detected: node at path %s references node at path %s", pathStr, prevPath)
+		}
+	}
+
+	p.visitedNodes[node] = append(paths, pathStr)
 	return nil
 }
+
+// areInDifferentServices checks if two paths are in different service definitions
+func areInDifferentServices(path1, path2 string) bool {
+	// Split paths into components
+	parts1 := strings.Split(path1, ".")
```

Comment:
nit: could use path.Parts() if you don't convert it to a string

Reply:
the function accepts strings and the field also has a list of strings `visitedNodes map[*yaml.Node][]string`, do you suggest to use path everywhere? or to convert the strings inside the func into paths and use `Parts()`? it looks like overhead, isn't it?

Reply:
some overhead indeed, but my point is that `Path` abstraction allows to ignore "implementation details" like the dot notation and `[]` or `*` patterns. I had in mind at some point we might want to use an alternate implementation vs `string` as we actually have to split this string many times during parsing.
From an encapsulation point of view, I'd prefer we add more utility func to `Path` to support required submatching. 
ANYWAY I'm fine we merge this PR as a quick fix for this issue, and revisit this later

Reply:
@ndeloof agree, that sounds reasonable.